### ISSUE TITLE
Llama3 8b bf16 cleanup error

### ISSUE THREAD

**Issue by salberdi-nvidia** [NVIDIA-NeMo/Megatron-Bridge#770]
**Describe the bug**

After finishing training, llama3 8b errors out with the following:
`[2025-09-24 16:08:38] iteration       25/      25 | consumed samples:         3200 | elapsed time per iteration (ms): 13029.8 | learning rate: 3.750000E-06 | global batch size:   128 | lm loss: 3.664289E+00 | loss scale: 1.0 | grad norm: 11.408 | number of skipped iterations:   0 | number of nan iterations:   0 |
[after training is done] datetime: 2025-09-24 16:08:38
[rank0]:[W924 16:08:40.642247764 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank1]:[W924 16:08:40.708499746 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank6]:[W924 16:08:40.728819510 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank7]:[W924 16:08:40.728828322 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank4]:[W924 16:08:40.736044290 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank2]:[W924 16:08:41.150081871 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank5]:[W924 16:08:41.193728099 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank3]:[W924 16:08:41.212050793 ProcessGroupNCCL.cpp:1505] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
[rank7]:[W924 16:08:42.579640379 observer.cpp:116] Warning: Leaked callback handle: 1 (function operator())
pure virtual method called
terminate called without an active exception
/nemo_run/scripts/pretrain_llama3_8b_bf16_gpus8_tp1_pp1_cp2_vpNone_ep1_mbs1_gbs128.sh: line 7: 4022568 Aborted                 bash -c 'numactl --cpunodebind=$((SLURM_LOCALID/4)) --membind=$((SLURM_LOCALID/4)) python /lustre/fsw/portfolios/general/users/salberdi/mbridge/llama3_mbridge/workloads/pretrain_llama3.1/Megatron-Bridge/scripts/performance/run_script.py --config_file /lustre/fsw/portfolios/general/users/salberdi/mbridge/llama3_mbridge/workloads/pretrain_llama3.1/Megatron-Bridge/scripts/performance/configs/llama3/llama3_8b_llm_pretrain.yaml --model_name llama3 --model_size 8b --compute_dtype bf16 --fp8_recipe ds --gpu h100 -a dummy -p dummy -ng 8 train.manual_gc=true train.manual_gc_interval=100 profiling.use_nsys_profiler=true profiling.profile_step_start=5 profiling.profile_step_end=6 profiling.profile_ranks=[0,1,2,3,4,5,6,7] profiling.record_shapes=false'
[rank1]:[W924 16:08:42.995442038 observer.cpp:116] Warning: Leaked callback handle: 1 (function operator())
pure virtual method called
terminate called without an active exception
[rank4]:[W924 16:08:43.094856710 observer.cpp:116] Warning: Leaked callback handle: 1 (function operator())
[rank6]:[W924 16:08:43.097780079 observer.cpp:116] Warning: Leaked callback handle: 1 (function operator())
pure virtual method called
terminate called without an active exception
pure virtual method called
terminate called without an active exception
Sep 24 16:08:43.179436 4020926 slurmstepd   0x15555186c640: error: *** STEP 5890589.1 ON cw-dfw-h100-004-344-012 CANCELLED AT 2025-09-24T16:08:43 ***`

**Steps/Code to reproduce bug**

Running llama3 8b with the default config on bf16

**Expected behavior**

Expected to run

**Additional context**

Using 25.09.rc8 container on H100

**Comment by salberdi-nvidia** [comment]
_target_: nemo_run.config.Script
args:
- --config_file
- /lustre/fsw/portfolios/general/users/salberdi/mbridge/llama3_mbridge/workloads/pretrain_llama3.1/Megatron-Bridge/scripts/performance/configs/llama3/llama3_8b_llm_pretrain.yaml
- --model_name
- llama3
- --model_size
- 8b
- --compute_dtype
- bf16
- --fp8_recipe
- ds
- --gpu
- h100
- -a
- dummy
- -p
- dummy
- -ng
- '8'
- train.manual_gc=true
- train.manual_gc_interval=100
- profiling.use_nsys_profiler=true
- profiling.profile_step_start=5
- profiling.profile_step_end=6
- profiling.profile_ranks=[0,1,2,3,4,5,6,7]
- profiling.record_shapes=false
entrypoint: python
env: {}
inline: ''
m: false
metadata: {}
path: /lustre/fsw/portfolios/general/users/salberdi/mbridge/llama3_mbridge/workloads/pretrain_llama3.1/Megatron-Bridge/scripts/performance/run_script.py

**Comment by ericharper** [comment]
@salberdi-nvidia , can you try #772 to see if it fixes the issue?

**Comment by salberdi-nvidia** [comment]
Still seeing this issue, affecting all llama models

**Comment by bdubauski** [comment]
The issue is still there with GA container 25.09.

**Comment by bdubauski** [comment]
@ananthsub @ericharper this is blocking the adoption of 25.09 by the upcoming LLMB release. Could you have someone revisit  the bug?

cc: @nv-mollys

**Comment by gautham-kollu** [comment]
Can we try cherry-picking https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/841 ?

**Comment by ko3n1g** [comment]
@gautham-kollu We still see this issue. Could you take a look?

**Comment by yaoyu-33** [comment]
Closing as part of issue cleanup — this issue has been stale with no recent activity. If you still need help, please open a new issue with updated details. Thank you!

**Comment by pruprakash** [comment]
## QA RCCA Coverage Analysis

**Issue Summary**: Llama3 8b bf16 cleanup error - distributed process group not destroyed after training

**Root Cause**: destroy_process_group() was not being called on normal training exit, leaving zombie ranks and NCCL communicators open.

**Fix**: PR #772 - Added destroy_process_group() call at end of performance script and training examples.

**Classification**: CODE BUG

**Test Coverage**: COVERED
- testcases/rcca/training/test_dist_cleanup.py
- test_normal_exit_destroys_process_group: Verifies destroy_process_group is called on normal training exit
- test_sigterm_exits_cleanly: Verifies clean exit on SIGTERM without NCCL errors

The test launches a minimal 2-rank torchrun job, runs 3 training steps, and asserts that destroy_process_group() is invoked before process exit.
